canals/pipeline/pipeline.py

```diff
@@ -420,57 +439,169 @@ def _check_max_loops(self, component_name: str):
+                    "Component '%s' is ready to run. All mandatory inputs received, all optional inputs are skipped.",
+                    name,
+                )
+                return "run"
+            if input_sockets == received_input_sockets | mandatory_input_sockets | skipped_optional_input_sockets:
+                # We received part of the optional inputs, the rest are skipped
+                logger.debug(
+                    "Component '%s' is ready to run. All mandatory inputs received, skipped some optional inputs: %s",
+                    name,
+                    skipped_optional_input_sockets,
+                )
+                return "run"
 
-        # Check if the expected inputs were all received
-        if self._check_received_vs_expected_inputs(name=name, inputs=inputs, expected_inputs=expected_inputs):
+        if set(input_components.values()).issubset(received_input_sockets):
+            # We have data from each connected input component.
+            # We reach this when the current component is the first of the pipeline or when
+            # it has defaults and all its input components have run.
+            logger.debug("Component '%s' is ready to run: all expected inputs were received.", name)
             return "run"
 
-        # This node is missing some inputs. Did all the upstream nodes run?
-        nodes_to_wait_for, _ = zip(*expected_inputs) if expected_inputs else ([], [])
-
-        # Some node upstream didn't run yet, so we should wait for them.
-        if not self._all_nodes_to_wait_for_run(nodes_to_wait_for=nodes_to_wait_for):
-            if not inputs_buffer:
-                # What if there are no components to wait for?
-                raise PipelineRuntimeError(
-                    f"'{name}' is stuck waiting for input, but there are no other components to run. "
-                    "This is likely a Canals bug. Open an issue at https://github.com/deepset-ai/canals."
-                )
+        ###############
+        # WAIT CHECKS #
+        ###############
+        if mandatory_input_sockets == received_input_sockets and skipped_optional_input_sockets.issubset(
+            optional_input_sockets
+        ):
+            # We received all of the inputs we need and some optional inputs have not been skipped
+            logger.debug(
+                "Component '%s' is waiting. All mandatory inputs received, some optional are not skipped: %s",
+                name,
+                optional_input_sockets - skipped_optional_input_sockets,
+            )
+            return "wait"
 
+        if any(self.graph.nodes[n]["visits"] == 0 for n in input_components.keys()):
+            # Some upstream component that must send input to the current component has yet to run.
+            # This happens when:
+            #   * Component has default values
+            #   * Component has some mandatory inputs
+            #   * Component has optional inputs that are not skipped
```

Comment:
This comment is confusing, those conditions don't make a component wait... What did you want to say?

Reply:
Yeah, it's a bit confusing. I'll update the comment to make it clearer.

Reply:
I decided to just remove those points.